### PR TITLE
Wip builder classes

### PR DESCRIPTION
**Pull request by JamesRezo**
This is an attempt to put the business logic of Satis into specific objects (Builders). The goal is to extend and/or reuse the business logic in other tools and maybe to simplify tests production.

### CODE REVIEW

File: src/Composer/Satis/Builder/PackagesBuilder.php [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+use Composer\Package\BasePackage;
+use Composer\Repository\PlatformRepository;
+use Composer\Package\Loader\ArrayLoader;
+use Composer\Package\Dumper\ArrayDumper;
+use Composer\DependencyResolver\Pool;
+use Composer\Json\JsonFile;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Builds the JSON stuff.
```
Comment (alcohol) [inline src/Composer/Satis/Builder/PackagesBuilder.php]: 'stuff' does not really qualify as a description..

File: src/Composer/Satis/Builder/DownloadsBuilder.php [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+use Symfony\Component\Console\Helper\HelperSet;
+use Composer\Factory;
+use Composer\IO\ConsoleIO;
+
+/**
+ * Builds the archives of the repository.
+ *
+ * @author James Hautot <james@rezo.net>
+ */
+class DownloadsBuilder extends Builder implements BuilderInterface
```
Comment (alcohol) [inline src/Composer/Satis/Builder/DownloadsBuilder.php]: Downloads is rather generic. Can you find something more specific to the purpose of this class?

File: src/Composer/Satis/Builder/Builder.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    /** @var OutputInterface $output The output Interface. */
+    protected $output;
+
+    /** @var string $outputDir The directory where to build. */
+    protected $outputDir;
+
+    /** @var array $config The parameters from ./satis.json. */
+    protected $config;
+
+    /** @var bool $skipErrors Escapes Exceptions if true. */
```
Comment (alcohol) [inline src/Composer/Satis/Builder/Builder.php]: Skips, not escapes.

File: src/Composer/Satis/Builder/Builder.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    protected $skipErrors;
+
+    /**
+     * Base Constructor.
+     *
+     * @param OutputInterface $output     The output Interface
+     * @param string          $outputDir  The directory where to build
+     * @param array           $config     The parameters from ./satis.json
+     * @param bool            $skipErrors Escapes Exceptions if true
```
Comment (alcohol) [inline src/Composer/Satis/Builder/Builder.php]: Same as previous comment.

File: src/Composer/Satis/Builder/DownloadsBuilder.php [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    /** @var HelperSet $helperSet A HelperSet instance. */
+    private $helperSet;
+
+    /**
+     * Builds the archives of the repository.
+     *
+     * @param array $packages List of packages to dump
+     */
+    public function dump(array $packages)
```
Comment (alcohol) [inline src/Composer/Satis/Builder/DownloadsBuilder.php]: This method still contains a **lot** of logic. It should just be dumping. Try to extract the white/black-listing and such into helper methods or even other classes possibly,

File: src/Composer/Satis/Builder/DownloadsBuilder.php (hunk 2 above)
Comment (JamesRezo) [inline src/Composer/Satis/Builder/DownloadsBuilder.php]: it cames from the original function name **dumpDownloads**. **ArchiveBuilder**, reffering to the config property, is perhaps a better name ?

File: src/Composer/Satis/Builder/PackagesBuilder.php (hunk 1 above)
Comment (JamesRezo) [inline src/Composer/Satis/Builder/PackagesBuilder.php]: Right. "Builds the JSON files" is better

File: src/Composer/Satis/Builder/DownloadsBuilder.php (hunk 5 above)
Comment (JamesRezo) [inline src/Composer/Satis/Builder/DownloadsBuilder.php]: I agree. It is also true for PackagesBuilder->select() 
I can work on both ...

Comment (alcohol) [inline src/Composer/Satis/Builder/DownloadsBuilder.php]: If you're willing to put in the time and work, that would be appreciated.

File: src/Composer/Satis/Builder/DownloadsBuilder.php (hunk 2 above)
Comment (alcohol) [inline src/Composer/Satis/Builder/DownloadsBuilder.php]: Was going to be my suggestion too yeah.

**Comment by JamesRezo** [comment]
Commit 84e99d61ad2a8ac5a0a4279fe1ebfbdf435139c2 (and the next...) provides a Helper Class for the renamed **ArchiveBuilder** class. And unit tests for this new object ;)

I am wondering now if it lowers the logic enough or not. For example, if it were useful to extract the pear-library test in a private doDump()-like method....

I also don't understand the presence of code after the effective dump. There are some setDistxxx methods applied to $package and, as far as i know, with no chance to throw an exception...

**Comment by JamesRezo** [comment]
Commit 2f4ec99745fd86232375f2d56bf73c1100acb9d2 disptach the logic of Packages selection with a few helper methods.
I can still work for other improvements until saturday. After that, I'll be away for three weeks ;)

File: src/Composer/Satis/Builder/ArchiveBuilderHelper.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * Tells if a package has to be dumped.
+     *
+     * @param  PackageInterface $package The package to be dumped
+     *
+     * @return bool false if the package has to be dumped.
+     */
+    public function skipDump(PackageInterface $package)
```
Comment (alcohol) [inline src/Composer/Satis/Builder/ArchiveBuilderHelper.php]: I would sooner opt for a name such as `isSkippable`, `isDumpable` (negating/inverse the result).

The function name should say something about the subject, `skipDump` relates more to the operator (the helper).

File: src/Composer/Satis/Builder/ArchiveBuilder.php [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+        /* @var \Composer\Downloader\DownloadManager $downloadManager */
+        $downloadManager = $factory->createDownloadManager($io, $composerConfig);
+
+        /* @var \Composer\Package\Archiver\ArchiveManager $archiveManager */
+        $archiveManager = $factory->createArchiveManager($composerConfig, $downloadManager);
+
+        $archiveManager->setOverwriteFiles(false);
+
+        shuffle($packages);
```
Comment (alcohol) [inline src/Composer/Satis/Builder/ArchiveBuilder.php]: Is there any reason for shuffling the packages?

Comment (JamesRezo) [inline src/Composer/Satis/Builder/ArchiveBuilder.php]: The purpose of this PR is just refactoring, not changing the behavior, so I just copied the code :)

shuffling was proposed by @joshdifabio and merged by @naderman in #225

Comment (alcohol) [inline src/Composer/Satis/Builder/ArchiveBuilder.php]: Ah, interesting.

File: src/Composer/Satis/Builder/ArchiveBuilderHelper.php (hunk 6 above)
Comment (JamesRezo) [inline src/Composer/Satis/Builder/ArchiveBuilderHelper.php]: isSkippable() is a quick fix, i'd prefer this one

**Comment by JamesRezo** [comment]
Well... I can't do more for now. Archive dump tests are a bit difficult to do in a short time. So are Package Selection tests.

@alcohol : Can we consider, nevertheless, that this proposal is sufficient?

**Comment by alcohol** [comment]
Well, I can YOLO it. How solid do you feel about your PR? Have you ran it yourself?

**Comment by JamesRezo** [comment]
I've ran it on two personnal projects and it seemed ok for me.
Truth is there are no tests in satis today to check for regressions... and as I said, i'm away for 3 weeks.
So... How do you feel ? :)

**Comment by alcohol** [comment]
@naderman how do you feel about this PR?

* [ ] happy
* [ ] sad
* [ ] indifferent

**Comment by naderman** [comment]
Thanks!

**Comment by Iruwen** [comment]
Stopped working for me after this merge, I'm now getting:
"PHP Fatal error:  Class 'Composer\Satis\Builder\Filesystem' not found in /var/www/satis/src/Composer/Satis/Builder/ArchiveBuilder.php on line 76"

**Comment by alcohol** [comment]
Should be fixed by b7fa484

**Comment by Iruwen** [comment]
That was fast. Works.

**Pull request merged**